runtime/string_functions.cpp

```diff
@@ -1977,9 +1977,27 @@ Optional<string> f$strstr(const string &haystack, const string &needle, bool bef
 string f$strtolower(const string &str) {
   int n = str.size();
 
+  // if there is no upper case char inside the string, we can
+  // return the argument unchanged, avoiding the allocation and data copying;
+  // while at it, memorize the first upper case char, so we can
+  // use memcpy to copy everything before that pos;
+  // note: do not use islower() here, the compiler does not inline that function call;
+  // it could be beneficial to use 256-byte LUT here, but SIMD approach could be even better
+  const char *end = str.c_str() + n;
+  const char *uppercase_pos = std::find_if(str.c_str(), end, [](char ch) {
+    return ch >= 'A' && ch <= 'Z';
+  });
+  if (uppercase_pos == end) {
+    return str;
+  }
+
   string res(n, false);
-  for (int i = 0; i < n; i++) {
-    res[i] = (char)tolower(str[i]);
+  int64_t lowercase_prefix = uppercase_pos - str.c_str();
+  if (lowercase_prefix != 0) { // avoid unnecessary function call
+    std::memcpy(res.buffer(), str.c_str(), lowercase_prefix);
+  }
+  for (int i = lowercase_prefix; i < n; i++) {
+    res[i] = static_cast<char>(tolower(str[i]));
```

Comment:
Passing `char` to `std::tolower` or similar methods is undefined behavior by according to the standard.
It must be converted to `unsigned char` manually.
You can check 'Notes' section [here](https://en.cppreference.com/w/cpp/string/byte/tolower).
```suggestion
    res[i] = static_cast<char>(std::tolower(static_cast<unsigned char>(str[i])));
```
Yeah, that's very weird and that's what C++ is :smiling_face_with_tear: